crates/locksmith/src/guard.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{common::GUARDS, error::LockType, tracker::GuardTracker};
+use parking_lot::{MutexGuard, RwLockReadGuard, RwLockWriteGuard};
+use snowflake::ProcessUniqueId;
+use std::{
+    borrow::{Borrow, BorrowMut},
+    ops::{Deref, DerefMut},
+};
+
+macro_rules! guard_struct {
+    ($HcGuard:ident, $Guard:ident, $lock_type:ident) => {
+        pub struct $HcGuard<'a, T: ?Sized> {
+            puid: ProcessUniqueId,
+            inner: Option<$Guard<'a, T>>,
+            fair_unlocking: bool,
+        }
+
+        impl<'a, T: ?Sized> $HcGuard<'a, T> {
+            pub fn new(inner: $Guard<'a, T>) -> Self {
+                let puid = ProcessUniqueId::new();
+                GUARDS
+                    .lock()
```

Comment:
I know we shouldn't go too crazy fixing potential problems in our code that is designed to fix other problems... but it's totally possible we could end up with a deadlock in this GUARDS mutex... do you think we should do a `try_lock_for(so_long_time_value)` that can just panic so we know there's a problem?